recipe-server/client/control/tests/components/recipes/test_GenericFormContainer.js

```diff
@@ -0,0 +1,102 @@
+/* eslint-disable react/prop-types, react/display-name, jsx-a11y/no-static-element-interactions */
+import React from 'react';
+import { mount } from 'enzyme';
+
+import GenericFormContainer from 'control/components/recipes/GenericFormContainer';
+
+describe('<GenericFormContainer>', () => {
+  const props = {
+    form: () => <div />,
+    formAction: async () => true,
+    onFailure: () => {},
+    onSuccess: () => {},
+    formProps: {},
+  };
+
+  it('should work', () => {
+    const wrapper = () => mount(<GenericFormContainer {...props} />);
+
+    expect(wrapper).not.toThrow();
+  });
+
+  it('should nest the `form` prop as a child', () => {
+    const CustomEl = () => <div id="test">Custom</div>;
+
+    const wrapper = mount(<GenericFormContainer {...props} form={CustomEl} />);
+
+    expect(wrapper.find('#test').length).toBe(1);
+    expect(wrapper.find('#test').text()).toContain('Custom');
+  });
+
+  it('should fire the formAction on submission', () => {
+    const CustomEl = ({ onSubmit }) => <div onClick={onSubmit}>Test</div>;
+    let fired = false;
+    const wrapper = mount(
+      <GenericFormContainer
+        {...props}
+        form={CustomEl}
+        formAction={() => {
+          fired = true;
+        }}
+      />,
+    );
+
+    wrapper.find(CustomEl).simulate('click');
+    expect(fired).toBe(true);
+  });
+
+  it('should handle the formAction failing', () => {
+    const CustomEl = ({ onSubmit }) => <div onClick={onSubmit}>Test</div>;
+    let failed = false;
+    const wrapper = mount(
+      <GenericFormContainer
+        {...props}
+        form={CustomEl}
+        onFailure={() => {
+          failed = true;
+        }}
+        formAction={() => {
+          throw new Error('Form action failed');
+        }}
+      />,
+    );
+
+    wrapper.find(CustomEl).simulate('click');
+    expect(failed).toBe(true);
+  });
+
+  it('should handle the formAction succeeding', () => {
+    const CustomEl = ({ onSubmit }) => <div onClick={onSubmit}>Test</div>;
+    let success = false;
+    const wrapper = mount(
+      <GenericFormContainer
+        {...props}
+        form={CustomEl}
+        onSuccess={() => {
+          success = true;
+        }}
+      />,
+    );
+
+    wrapper.find(CustomEl).simulate('click');
+    // formAction is async so we need to wait for that promise to resolve.
+    setTimeout(() => {
```

Comment:
setTimeout shouldn't be used in tests like this. The onSuccess callback should resolve a promise that this function can await on. Also, since this test is not async, the test will pass before this callback fires.